src/uct/ib/rc/verbs/rc_verbs_ep.c

```diff
@@ -596,6 +596,38 @@ ucs_status_t uct_rc_verbs_ep_get_address(uct_ep_h tl_ep, uct_ep_addr_t *addr)
     return UCS_OK;
 }
 
+int uct_rc_verbs_ep_is_connected(const uct_ep_h tl_ep,
+                                 const uct_ep_is_connected_params_t *params)
+{
+    uct_rc_verbs_ep_t *ep = ucs_derived_of(tl_ep, uct_rc_verbs_ep_t);
+    const uct_rc_verbs_ep_addr_t *rc_addr;
+    uint32_t addr_qp;
+    ucs_status_t status;
+    struct ibv_ah_attr ah;
+    uint32_t qp_num;
+    const uct_ib_address_t *ib_addr;
+
+    if (!ucs_test_all_flags(params->field_mask,
+            UCT_EP_IS_CONNECTED_FIELD_EP_ADDR |
+            UCT_EP_IS_CONNECTED_FIELD_DEVICE_ADDR)) {
+        ucs_error("missing params (field_mask: %lu), both device_addr and "
+                      "ep_addr must be provided.", params->field_mask);
+        return 0;
+    }
+
+    rc_addr = (const uct_rc_verbs_ep_addr_t*)params->ep_addr;
+    addr_qp = uct_ib_unpack_uint24(rc_addr->qp_num);
+
+    status = uct_ib_query_qp_peer_info(ep->qp, &ah, &qp_num);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    ib_addr = (const uct_ib_address_t*)params->device_addr;
+    return uct_ib_iface_is_same_device(ib_addr, ah.dlid, &ah.grh.dgid) &&
+           (qp_num == addr_qp);
```

Comment:
we can probably compare qp_num first.
btw, can we share code between rc_verbs and rc_mlx5? maybe some common function that receives destination ah_attr and qp_num as parameter

Reply:
the only common code is:
```
    if (params->field_mask & UCT_EP_IS_CONNECTED_FIELD_EP_ADDR) {
        if (qp_num != addr_qp) {
            return 0;
        }
    }

    ib_addr = (const uct_ib_address_t*)params->device_addr;
    return uct_ib_iface_is_same_device(ib_addr, ah_attr.dlid, &ah_attr.grh.dgid);
}
```
and even the if we should check outside in order to pass addr_qp, so not much code left.